prboom2/src/dsda/hud_components/weapon_text.c

```diff
@@ -31,57 +32,106 @@ static void dsda_UpdateComponentText(char* str, size_t max_size) {
 
   player = &players[displayplayer];
 
-  if (local->grid)
-    snprintf(
-      str,
-      max_size,
-      "%sW %s%c %s%c %c\n"
-      "%sP %s%c %c %c\n"
-      "%sN %s%c %c %c",
-      dsda_TextColor(dsda_tc_exhud_weapon_label),
-      player->powers[pw_strength] ? dsda_TextColor(dsda_tc_exhud_weapon_berserk) :
-                                    dsda_TextColor(dsda_tc_exhud_weapon_owned),
-      player->weaponowned[0] ? '1' : ' ',
-      dsda_TextColor(dsda_tc_exhud_weapon_owned),
-      player->weaponowned[1] ? '2' : ' ',
-      player->weaponowned[2] ? '3' : ' ',
-      dsda_TextColor(dsda_tc_exhud_weapon_label),
-      dsda_TextColor(dsda_tc_exhud_weapon_owned),
-      player->weaponowned[3] ? '4' : ' ',
-      player->weaponowned[4] ? '5' : ' ',
-      player->weaponowned[5] ? '6' : ' ',
-      dsda_TextColor(dsda_tc_exhud_weapon_label),
-      dsda_TextColor(dsda_tc_exhud_weapon_owned),
-      player->weaponowned[6] ? '7' : ' ',
-      player->weaponowned[7] ? '8' : ' ',
-      player->weaponowned[8] ? '9' : ' '
-    );
-  else
-    snprintf(
-      str,
-      max_size,
-      "%sWPN %s%c %s%c %c %c %c %c %c %c %c",
-      dsda_TextColor(dsda_tc_exhud_weapon_label),
-      player->powers[pw_strength] ? dsda_TextColor(dsda_tc_exhud_weapon_berserk) :
-                                    dsda_TextColor(dsda_tc_exhud_weapon_owned),
-      player->weaponowned[0] ? '1' : ' ',
-      dsda_TextColor(dsda_tc_exhud_weapon_owned),
-      player->weaponowned[1] ? '2' : ' ',
-      player->weaponowned[2] ? '3' : ' ',
-      player->weaponowned[3] ? '4' : ' ',
-      player->weaponowned[4] ? '5' : ' ',
-      player->weaponowned[5] ? '6' : ' ',
-      player->weaponowned[6] ? '7' : ' ',
-      player->weaponowned[7] ? '8' : ' ',
-      player->weaponowned[8] ? '9' : ' '
-    );
+  if (local->show_wpn) {
+    if (local->grid)
+      snprintf(
+        str,
+        max_size,
+        "%sW %s%c %s%c %c\n"
```

Comment:
There are some options here to make it cleaner.

We could have a `const char* format;` in the component struct, define 4 static format strings, and set the `format` when parsing the args in the init function. Then you won't need the `show_wpn` attribute and also don't need to duplicate the interpolation calls. The downside is the format string is no longer visible directly where we are interpolating.

Another option is to set `label_w`, `label_p`, `label_n`, and `label_wpn` when initializing (either being, e.g., `"W "` or `""`) and interpolating those labels into the existing format strings (see `composite_time.c`). That has the benefit of keeping the format strings visible in the print functions, but has the disadvantage of adding even more interpolations to them.

I would probably go for the second option.